launch_ros/launch_ros/descriptions/node.py

```diff
@@ -0,0 +1,327 @@
+            self.__expanded_parameter_arguments.append((param_file_path, True))
+            cmd_ext = ['--params-file', f'{param_file_path}']
+            cmd.extend([normalize_to_list_of_substitutions(x) for x in cmd_ext])
+            assert os.path.isfile(param_file_path)
+        # expand parameters too
+        if self.__parameters is not None:
+            evaluated_parameters = evaluate_parameters(context, self.__parameters)
+            for params in evaluated_parameters:
+                is_file = False
+                if isinstance(params, dict):
+                    param_argument = self._create_params_file_from_dict(params)
+                    is_file = True
+                    assert os.path.isfile(param_argument)
+                elif isinstance(params, pathlib.Path):
+                    param_argument = str(params)
+                    is_file = True
+                elif isinstance(params, Parameter):
+                    param_argument = self._get_parameter_rule(params, context)
+                else:
+                    raise RuntimeError('invalid normalized parameters {}'.format(repr(params)))
+                if is_file and not os.path.isfile(param_argument):
+                    self.__logger.warning(
+                        'Parameter file path is not a file: {}'.format(param_argument),
+                    )
+                    continue
+                self.__expanded_parameter_arguments.append((param_argument, is_file))
+                cmd_ext = ['--params-file' if is_file else '-p', f'{param_argument}']
+                cmd.extend([normalize_to_list_of_substitutions(x) for x in cmd_ext])
+        # expand remappings too
+        global_remaps = context.launch_configurations.get('ros_remaps', None)
+        if global_remaps or self.__remappings:
+            self.__expanded_remappings = []
+        if global_remaps:
+            self.__expanded_remappings.extend(global_remaps)
+        if self.__remappings:
+            self.__expanded_remappings.extend([
+                (perform_substitutions(context, src), perform_substitutions(context, dst))
+                for src, dst in self.__remappings
+            ])
+        if self.__expanded_remappings:
+            cmd_ext = []
+            for src, dst in self.__expanded_remappings:
+                cmd_ext.extend(['-r', f'{src}:={dst}'])
+            cmd.extend([normalize_to_list_of_substitutions(x) for x in cmd_ext])
+        # Prepare the ros_specific_arguments list and add it to the context so that the
+        # LocalSubstitution placeholders added to the the cmd can be expanded using the contents.
+        ros_specific_arguments: Dict[str, Union[str, List[str]]] = {}
+        if self.__node_name is not None:
+            ros_specific_arguments['name'] = '__node:={}'.format(self.__expanded_node_name)
+        if self.__expanded_node_namespace != '':
+            ros_specific_arguments['ns'] = '__ns:={}'.format(self.__expanded_node_namespace)
```

Comment:
@roger-strain meta: this will probably need a revisit. A blanket `__node:=` remapping will apply to all nodes in the same executable. Probably not what a user would intend in this case. Same for `__ns:=`.

Reply:
This is probably a carry-over from the earlier design which assumed one `Node` per executable process. I'm looking at the docs here (https://docs.ros.org/en/foxy/Guides/Node-arguments.html) and it looks like I need to prefix things with the "old" name of the node. That's something that I don't believe we've captured in the `descriptions.Node` structure to this point. I'm adding a new parameter to capture `original_node_name`; if it's not set, there's nothing we can do but let the global apply. (This will actually work out okay, though, because I think it maintains current functionality for backward-compatibility.)

I'm now having another thought, though, which is related to this. For "normal" (read: Not node name/namespace) parameters, the docs show the ability to prefix individual parameters with the (new) node name. I'm guessing this is something we'll need to bake in at this layer as well? How does that come into play if we're passed or creating parameter dictionary files? Not sure how far down this rabbit hole is the responsibility of this layer, vs telling the developer they need to specify things with the appropriate prefix. Would love your thoughts on which parts to tackle internally.

Reply:
I'll start by saying that some of these rough edges when dealing with remapping predate your work, and that I'm not pushing to get them all addressed here.

I don't think `launch` should be mutating parameter files to propagate a remapping and such. That's on the underlying executable's plate at best. IMHO `launch` should simply provide a way to express the intent to remap. 

For executables that bear many nodes, the current command line interface allows both node-specific and blanket configurations. To support the latter, we can simply let the `ROSExecutable` action take extra command line arguments. To support the former, `Node` descriptions have to bear the original name, that's right. We can allow a single `Node` description to not specify it for convenience, in which case I think the `ROSExecutable` action should enforce only one such description is given.

Reply:
I think I ended up trying to make this more complicated than it needed to be. If we're setting aside the dictionary files, then in all cases we know what the applied node name should be by the time we're setting up command line remaps, so I can just add that as the normal way of doing it.

Here's a question on parameters, though. Is there such a thing as a node-specific (non-remapping) parameter? If so, I don't see it in the sample documentation I've been referring to.